File: python/basics/arrays/345_reverse_vowels_of_string.py
'''
https://leetcode.com/problems/reverse-vowels-of-a-string/
'''

def reverse_vowels(s: str) -> str:
    vowels = set("aeiouAEIOU")
    s = list(s)
    N=len(s)
    l = 0
    r = len(s) -1
    print(s)
    while l < r:
        while l < r and s[l] not in vowels:
            l += 1
        while l < r and s[r] not in vowels:
            r -= 1
        print(l)
        print(r)
        if l < r and s[l] in vowels and s[r] in vowels:
            s[l],s[r] = s[r], s[l]
            l += 1
            r -= 1

    return "".join(s)



s = "IceCreAm"
#Output: "AceCreIm"
result = reverse_vowels(s)
print("result: " + result)

s = "leetcode"
#Output: "leotcede"
result = reverse_vowels(s)
print("result: " + result)

File: problem_solving/arrays/cpp/345_reverse_vowels_of_string.cpp
/*
'''
https://leetcode.com/problems/reverse-vowels-of-a-string/
'''
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
using namespace std;


string reverse_vowels(std::string s) {
    std::string result;
    int l = 0;
    std::unordered_set<char> vowels = {'a','e','i','o','u', 'A','E','I','O','U'};
    int r = s.length() - 1;
    while(l < r) {
        while(l < r && vowels.find(s[l]) == vowels.end() ) {
            l++;
        }
        while(l < r && vowels.find(s[r]) == vowels.end() ) {
            r--;
        }
        if(l < r) {
            std::swap(s[l], s[r]);
            l++;
            r--;
        }
    }
    return s;
}



int main() {
    std::string s = "IceCreAm";
    //#Output: "AceCreIm"
    std::string result = reverse_vowels(s);
    std::cout << result << std::endl;

    s = "leetcode";
    //#Output: "leotcede"
    result = reverse_vowels(s);
    std::cout << result << std::endl;
}